UPGRADE_GUIDE.md

```diff
@@ -2,8 +2,133 @@
 
 Use the following guide to assist in the upgrade process of the `easypost-ruby` library between major versions.
 
+- [Upgrading from 4.x to 5.0](#upgrading-from-4x-to-50)
+- [Upgrading from 3.x to 4.0](#upgrading-from-3x-to-40)
+
+## Upgrading from 4.x to 5.0
+
+### 5.0 High Impact Changes
+
+- [Thread-Safe With Client Object](#50-thread-safe-with-client-object)
+- [Updated Dependencies](#50-updated-dependencies)
+- [Improved Error Handling](#50-improved-error-handling)
+
+### 5.0 Medium Impact Changes
+
+- [Corrected Naming Conventions](#50-corrected-naming-conventions)
+
+### 5.0 Low Impact Changes
+
+- [Beta Namespace Changed](#50-beta-namespace-changed)
+
+## 5.0 Thread-Safe with Client Object
+
+Likelihood of Impact: High
+
+This library is now thread-safe with the introduction of a new `Client` object. Instead of defining a global API key that all requests use, you now create an `Client` object and pass your API key to it with optional open and read timeout params. You then call your desired functions against a `service` which coincide with the `Client` object:
+
+```ruby
+# Old way
+require 'easypost'
+
+EasyPost.api_key = ENV['EASYPOST_API_KEY']
+shipment = EasyPost::Shipment.retrieve('shp_...')
+
+# New way
+require 'easypost'
+
+client = EasyPost::Client.new(api_key: ENV['EASYPOST_TEST_API_KEY'])
+shipment = client.shipment.retrieve('shp_...')
+```
+
+All instance methods are now static with the exception of `lowest_rate` as these make API calls and require the EasyPostClient (EasyPost objects do not contain an API key to make API calls with).
+
+Previously used `.save()` instance methods are now static `.update()` functions where you specify first the ID of the object you are updating and second, the parameters that need updating.
+
+Functions no longer accept an API key as an optional parameter. If you need per-function API key changes, create a new EasyPostClient object and call the function on the new client that uses the API key you need.
```

Comment:
question: was this ever possible in ruby? If not, let's not include it.

Reply:
Idk why it's not showing, but usually when you hover over a comment on github it'll show which lines I selected. The one I'm talking about is passing an API key per function. I don't believe that was ever a feature of Ruby

Reply:
This was copied and pasted from the Python upgrade guide. The sentence is telling users if they want to use a different API key for a request, they have to create another `Client` to make that happen, which is possible in all the thread-safe CLs

Reply:
Right, but in Python and a couple other libs, you could pass an API key directly to functions. What I'm suggesting is that we can't just copy/paste from lib to lib because they are all different. Ruby (to my remembrance), never allowed you to pass an API key directly to a function so this statement is incorrect as written for this library and could be confusing to users. I'm suggesting we remove it completely because it's not relevant to this library. We describe how to pass API keys up in the client section.